api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -2,38 +2,236 @@ package messagebuffer
 
 import (
 	"encoding/hex"
+	"fmt"
+	"io"
 	"sync"
+	"time"
 
+	rbt "github.com/emirpasic/gods/trees/redblacktree"
+	"github.com/nknorg/nkn/v2/config"
 	"github.com/nknorg/nkn/v2/pb"
 )
 
+const maxSeqId int64 = 1000000
+
+type sequence struct {
+	sync.RWMutex
+	seqId int64
+}
+
+// msg node with create time and expire time
+type msgNode struct {
+	msg      *pb.Relay
+	createId int64
+	expireId int64
+}
+
+func (m *msgNode) GetSize() int {
+	// m.createId(int64), m.expireId(int64), m.msg.MaxHoldingSeconds(int32), m.msg.SigChainLen(int32)
+	return (8 + 8 + 4 + 4 + len(m.msg.SrcIdentifier) + len(m.msg.SrcPubkey) + len(m.msg.DestId) +
+		len(m.msg.Payload) + len(m.msg.BlockHash) + len(m.msg.LastHash))
+}
+
+type indexNode struct {
+	clientID string
+	createId int64
+	expireId int64
+}
+
 // MessageBuffer is the buffer to hold message for clients not online
 type MessageBuffer struct {
 	sync.Mutex
-	buffer map[string][]*pb.Relay
+	buffer       map[string]*rbt.Tree // map ClinetId to msgs
+	totalBufSize int                  // the buffer size in bytes
+	treeCreate   *rbt.Tree            // index tree of create time
+	treeExpire   *rbt.Tree            // index tree of expire time
+	seq          sequence
+}
+
+// comparation function of red and black tree
+func compare(a, b interface{}) (res int) {
+	aInt := a.(int64)
+	bInt := b.(int64)
+	switch {
+	case aInt < bInt:
+		res = -1
+	case aInt == bInt:
+		res = 0
+	case aInt > bInt:
+		res = 1
+	}
+	return
 }
 
 // NewMessageBuffer creates a MessageBuffer
 func NewMessageBuffer() *MessageBuffer {
-	return &MessageBuffer{
-		buffer: make(map[string][]*pb.Relay),
+
+	msgBuffer := &MessageBuffer{
+		buffer:     make(map[string]*rbt.Tree), // *pb.Relay
+		treeCreate: rbt.NewWith(compare),
+		treeExpire: rbt.NewWith(compare),
+	}
+
+	go msgBuffer.BufferMonitor()
+
+	return msgBuffer
+}
+
+// get next sequence id
+func (messageBuffer *MessageBuffer) getNextSeq() int64 {
+
+	messageBuffer.seq.Lock()
+	defer messageBuffer.seq.Unlock()
+	messageBuffer.seq.seqId++
+
+	if messageBuffer.seq.seqId >= maxSeqId {
+		messageBuffer.seq.seqId = 0
 	}
+
+	return messageBuffer.seq.seqId
 }
 
 // AddMessage adds a message to message buffer
 func (messageBuffer *MessageBuffer) AddMessage(clientID []byte, msg *pb.Relay) {
+
+	if msg.MaxHoldingSeconds <= 0 { // no need buffer
+		return
+	}
+
+	now := time.Now().UnixMilli()
+	seqId := messageBuffer.getNextSeq()
```

Comment:
we can remove lock in messageBuffer.seq if we move L114-115 to L100

Reply:
yes